File: src/utils/agentic/routing.ts
/**
 * Agentic Workflow Routing & Orchestration
 * Deterministic routing rules and state machine logic
 */

import type { AgentState, RunContext, RoutingRule, AgentRegistry } from "@/types/agentic";

/**
 * UI Route mapping - frontend routes reflect agent state
 */
export const UI_ROUTES: Record<AgentState, string> = {
  CREATED: "/agentic/upload",
  REGISTERED: "/agentic/analyze",
  ANALYZED: "/agentic/proposal",
  PROPOSED: "/agentic/review",
  APPROVED: "/agentic/run",
  REJECTED: "/agentic/review",
  EXECUTING: "/agentic/progress",
  COMPLETED: "/agentic/results",
  FAILED: "/agentic/error",
  STOPPED: "/agentic/progress",
};

/**
 * Routing table - deterministic state transitions
 */
export const ROUTING_TABLE: RoutingRule[] = [
  {
    from: "CREATED",
    to: "REGISTERED",
    // Auto-transition on file upload
  },
  {
    from: "REGISTERED",
    to: "ANALYZED",
    agent: "analyzer-agent",
  },
  {
    from: "ANALYZED",
    to: "PROPOSED",
    agent: "planner-agent",
  },
  {
    from: "PROPOSED",
    to: "APPROVED",
    requiresHuman: true,
  },
  {
    from: "APPROVED",
    to: "EXECUTING",
    agent: "executor-agent",
  },
  {
    from: "EXECUTING",
    to: "COMPLETED",
    // Auto-transition on completion
  },
];

/**
 * Agent registry - capabilities-based routing
 */
export const AGENT_REGISTRY: AgentRegistry = {
  "analyzer-agent": {
    inputs: ["image", "video"],
    outputs: ["analysis"],
    costProfile: "low",
  },
  "planner-agent": {
    inputs: ["analysis"],
    outputs: ["proposal"],
    costProfile: "low",
  },
  "executor-agent": {
    inputs: ["proposal"],
    outputs: ["assets"],
    costProfile: "high",
  },
};

/**
 * Find routing rule for current state
 */
export function findRoutingRule(state: AgentState): RoutingRule | null {
  return ROUTING_TABLE.find((rule) => rule.from === state) || null;
}

/**
 * Check if state requires human approval
 */
export function requiresHumanApproval(state: AgentState): boolean {
  const rule = findRoutingRule(state);
  return rule?.requiresHuman === true;
}

/**
 * Get next state for context
 */
export function getNextState(ctx: RunContext): AgentState | null {
  const rule = findRoutingRule(ctx.state);
  if (!rule) return null;

  // Check if human approval is required
  if (rule.requiresHuman && !ctx.humanApproved) {
    return ctx.state; // Stay in current state
  }

  return rule.to;
}

/**
 * Check if state transition is allowed
 */
export function canTransition(from: AgentState, to: AgentState): boolean {
  const rule = ROUTING_TABLE.find((r) => r.from === from && r.to === to);
  return rule !== undefined;
}

/**
 * Get UI route for state
 */
export function getRouteForState(state: AgentState): string {
  return UI_ROUTES[state] || "/agentic";
}

File: backend/app/orchestration/routing.py
"""
Agent Routing Table
Deterministic routing rules for multi-agent workflow.
"""

from enum import Enum
from typing import Optional, Dict, Any, List


class AgentRole(str, Enum):
    """Agent role types in the workflow."""
    PLANNER = "planner"
    CRITIC = "critic"
    EXECUTOR = "executor"


class WorkflowState(str, Enum):
    """Workflow state enumeration."""
    CREATED = "CREATED"
    PLANNED = "PLANNED"
    CRITIQUED = "CRITIQUED"
    APPROVED = "APPROVED"
    EXECUTING = "EXECUTING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


# Routing table defining state transitions and agent assignments
ROUTING_TABLE: List[Dict[str, Any]] = [
    {
        "from": WorkflowState.CREATED,
        "to": WorkflowState.PLANNED,
        "role": AgentRole.PLANNER,
        "requires_human": False,
    },
    {
        "from": WorkflowState.PLANNED,
        "to": WorkflowState.CRITIQUED,
        "role": AgentRole.CRITIC,
        "requires_human": False,
    },
    {
        "from": WorkflowState.CRITIQUED,
        "to": WorkflowState.APPROVED,
        "role": None,
        "requires_human": True,  # Human-in-the-loop approval
    },
    {
        "from": WorkflowState.APPROVED,
        "to": WorkflowState.EXECUTING,
        "role": AgentRole.EXECUTOR,
        "requires_human": False,
    },
    {
        "from": WorkflowState.EXECUTING,
        "to": WorkflowState.COMPLETED,
        "role": None,
        "requires_human": False,
    },
]


def get_next_state(current_state: WorkflowState) -> Optional[Dict[str, Any]]:
    """Get the next routing rule for a given state."""
    for rule in ROUTING_TABLE:
        if rule["from"] == current_state:
            return rule
    return None


def requires_human_approval(state: WorkflowState) -> bool:
    """Check if a state requires human approval."""
    rule = get_next_state(state)
    return rule.get("requires_human", False) if rule else False

